### PR TITLE
Directives & Pipes task

### PR DESCRIPTION
**Pull request by GlebZhidovich**
1. https://github.com/rolling-scopes-school/tasks/blob/master/tasks/angular/components-directives-pipes.md

2. Main layout is implemented (+10)
 Implemented a fake search functionality: by submitting the form on the Header component, search results are shown (+20)
 The Search result item component contains all the necessary data (+10)
 The Filtering criteria block toggle functionality is implemented (+5)
 The colorful border under the Search result item is implemented (+15)
 Search result sorting is implemented (Should work both in the direction of decreasing values, and in the direction of increasing values) (+20)
 Search result filtering by key words is implemented (+20)

### CODE REVIEW

File: youtube-client-app/.browserslistrc [hunk 1]
```diff
@@ -0,0 +1,18 @@
+# This file is used by the build system to adjust CSS and JS output to support the specified browsers below.
```
Comment (sergey496) [inline youtube-client-app/.browserslistrc:1]: test comment

File: youtube-client-app/src/app/app/app.component.html [hunk 2]
```diff
@@ -0,0 +1,2 @@
+<app-header (getVideoName)="getName($event)" (sb)="sortBy($event)" (sbw)="sortByWord($event)"></app-header>
```
Comment (sergey496) [inline youtube-client-app/src/app/app/app.component.html:1]: При более 2 аттрибутов желательно их переносить на новую строку
<app-header 
     (getVideoName)="getName($event)" 
     (sb)="sortBy($event)" 
     (sbw)="sortByWord($event)">
</app-header>

File: youtube-client-app/src/app/app/app.component.ts [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+  templateUrl: './app.component.html',
+  styleUrls: ['./app.component.scss'],
+})
+export class AppComponent {
+  public videoName: string;
+  public sort: string;
+  public sortWord: string;
+
+  public getName(name: string): void {
+  this.videoName = name;
```
Comment (sergey496) [inline youtube-client-app/src/app/app/app.component.ts:14]: добавь отступы

File: youtube-client-app/src/app/components/header/header.component.html [hunk 4]
```diff
@@ -0,0 +1,5 @@
+<div class="header">
+  <div class="header-item"></div>
```
Comment (sergey496) [inline youtube-client-app/src/app/components/header/header.component.html:2]: пустой див нужен?

File: youtube-client-app/src/app/components/header/header.component.ts [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+@Component({
+  selector: 'app-header',
+  templateUrl: './header.component.html',
+  styleUrls: ['./header.component.scss'],
+  changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class HeaderComponent {
+
+  @Output() public getVideoName: EventEmitter<string> = new EventEmitter<string>();
+  @Output() public sb: EventEmitter<string> = new EventEmitter<string>();
```
Comment (sergey496) [inline youtube-client-app/src/app/components/header/header.component.ts:12]: strange name **sb**

File: youtube-client-app/src/app/components/header/header.component.ts [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+  @Output() public sbw: EventEmitter<string> = new EventEmitter<string>();
+
+  constructor(private cdr: ChangeDetectorRef) {
```
Comment (sergey496) [inline youtube-client-app/src/app/components/header/header.component.ts:15]: do you need here **ChangeDetectedRef**?

File: youtube-client-app/src/app/components/header/search-panel/search-panel.component.html [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+
+    <button class="search-btn" (click)="searchVideo()" mat-raised-button>Search</button>
+
+    <div class="setting-btn" (click)="isFilterShow()">
+      <img src="assets/img/search_settings.svg" alt="settings">
+    </div>
+
+  </div>
+
+  <app-sort-panel [@open-close] *ngIf="isFilter"> (sb)="sortBy($event)" (sbw)="sortByWord($event)"></app-sort-panel>
```
Comment (sergey496) [inline youtube-client-app/src/app/components/header/search-panel/search-panel.component.html]: **(sb)="sortBy($event)" (sbw)="sortByWord($event)">**  is it correct?

File: youtube-client-app/src/app/components/header/search-panel/search-panel.component.ts [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+
+  public searchVideo(): void {
+  const elem: HTMLInputElement = this.input.nativeElement;
+  const value: string = elem.value;
+  this.getName.emit(value);
+  elem.value = '';
+  }
+
+  public sortBy(type: string): void {
+	this.sb.emit(type);
```
Comment (sergey496) [inline youtube-client-app/src/app/components/header/search-panel/search-panel.component.ts:47]: please align the same in all methods

File: youtube-client-app/src/app/components/header/sort-panel/sort-panel.component.ts [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+import { ChangeDetectionStrategy, Component, EventEmitter, Output } from '@angular/core';
+
+@Component({
+  selector: 'app-sort-panel',
+  templateUrl: './sort-panel.component.html',
+  styleUrls: ['./sort-panel.component.scss'],
+  changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class SortPanelComponent {
+  private date: string = 'Up';
```
Comment (sergey496) [inline youtube-client-app/src/app/components/header/sort-panel/sort-panel.component.ts:10]: please create a new type for "up" and "down"

File: youtube-client-app/src/app/components/search-item/search-item.component.ts [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+  for (const className in border) {
+  if (border[className]) { return className; }
+  }
+  }
+
+  public ngOnInit(): void {
+	this.statistics = this.videoData.statistics;
+	this.snippet = this.videoData.snippet;
+  this.day = new Date(this.snippet.publishedAt).valueOf();
+  console.log(this.day);
```
Comment (sergey496) [inline youtube-client-app/src/app/components/search-item/search-item.component.ts]: please remove

Comment (sergey496) [reply youtube-client-app/src/app/components/search-item/search-item.component.ts]: and align

File: youtube-client-app/src/app/components/search-results/search-results.component.ts [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+  selector: 'app-search-results',
+  templateUrl: './search-results.component.html',
+  styleUrls: ['./search-results.component.scss'],
+  providers: [SortService],
+  changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class SearchResultsComponent implements OnChanges {
+  public videosData: IVideo[];
+
+  constructor(private ss: SortService) {
```
Comment (sergey496) [inline youtube-client-app/src/app/components/search-results/search-results.component.ts:29]: **ss** is bad name, I like **private sortService:SortService**

File: youtube-client-app/src/app/components/search-results/search-results.component.ts [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+  if (name) {
+  this.videosData = this.ss.sort(name, this.videosData);
+  }
+  }
+
+  public ngOnChanges(changes: SimpleChanges): void {
+  const data: [string, SimpleChange][] = Object.entries(changes);
+  data.forEach((arr: [string, SimpleChange]): void => {
+  const [name, data]: [string, SimpleChange] = arr;
+  if (name === 'sort') {
```
Comment (sergey496) [inline youtube-client-app/src/app/components/search-results/search-results.component.ts:50]: add new type for 'sort'

File: youtube-client-app/src/app/components/search-results/sort-by-name.pipe.ts [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+import { IVideo } from '../../models/search-response.model';
+
+@Pipe({
+  name: 'sortByName',
+})
+export class SortByNamePipe implements PipeTransform {
+
+  public transform(value: IVideo[], ...args: string[] | undefined[]): IVideo[] {
+  if (value) {
+	const [name]: string[] | undefined[] = args;
```
Comment (sergey496) [inline youtube-client-app/src/app/components/search-results/sort-by-name.pipe.ts:11]: may be better (string | undefined)[] or Array<string | undefined>

File: youtube-client-app/src/app/components/header/header.component.html (hunk 4 above)
Comment (GlebZhidovich) [reply youtube-client-app/src/app/components/header/header.component.html:2]: ну тут я его добавил чтобы разделить на три части header, может подскажешь как по другому можно?

File: youtube-client-app/src/app/components/search-results/search-results.component.ts (hunk 12 above)
Comment (GlebZhidovich) [reply youtube-client-app/src/app/components/search-results/search-results.component.ts:50]: тут не понял

Comment (sergey496) [reply youtube-client-app/src/app/components/search-results/search-results.component.ts:50]: создай enum со значениями sort = 'sort', videoName= 'videoName' и используй при для проверки **name ===<nameEnum>.sort**

**Pull request merged**